src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -792,52 +803,97 @@ static int parse_imgpath(const std::string &imgpath)
   return 0;
 }
 
-static int do_list_mapped_devices()
+static int get_mapped_info(int pid)
 {
-  char path[64];
-  int m = 0;
-  int fd[2];
+  int r;
+  std::string path = "/proc/" + stringify(pid) + "/cmdline";
+  std::ifstream ifs;
+  std::string cmdline;
+  std::vector<const char*> args;
 
-  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
-    int r = -errno;
-    cerr << "rbd-nbd: socketpair failed: " << cpp_strerror(-r) << std::endl;
-    return r;
+  ifs.open(path.c_str(), std::ifstream::in);
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for (unsigned int index = 0; index < cmdline.size(); index++) {
+    if (((unsigned char)cmdline[index]) < ' ') {
+      cmdline[index] = ' ';
+      index++;
+    }
   }
 
-  while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
-      break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
+  char *pch = const_cast<char*>(cmdline.c_str());
+  pch = strtok(pch, " ");
+  if (strstr(pch, "rbd-nbd") != NULL) { 
+    pch = strtok(NULL, " ");
+  } else {
+    return EXIT_FAILURE;
   }
 
-  close(fd[0]);
-  close(fd[1]);
+  while (pch != NULL) {
+    args.push_back(pch);
+    pch = strtok(NULL, " ");
+  }
```

Comment:
I don't think this manual parsing is needed. What I meant was to take `cmdline`, which contains arguments separated by '\0', and split it into `args`, using something like below (this is just a not tested example, may be you will find a better way):
```
   for (int i = 0; i < cmdline.size(); i++) {
      args.push_back(&cmdline[i]);
      while(cmdline[i] != '\0') {
        i++;
      }
   }
```
Then you just pass `args` to `parse_args()`. And your `parse_args()` should do `md_config_t().parse_argv(args)` at start.